Simbody/src/CableSpan.cpp

```diff
@@ -0,0 +1,3028 @@
+        for (CurveSegment& curve : m_curveSegments) {
+            curve.setSubsystem(subsystem);
+        }
+    }
+
+    CableSubsystem& updSubsystem()
+    {
+        SimTK_ERRCHK_ALWAYS(
+            m_subsystem,
+            "CableSpan::Impl::updSubsystem()",
+            "CableSpan not yet adopted by any CableSubsystem");
+        return *m_subsystem;
+    }
+
+    // Get the index of this CableSpan in the CableSubsystem.
+    CableSpanIndex getIndex() const
+    {
+        SimTK_ERRCHK1_ALWAYS(
+            m_index.isValid(),
+            "CableSpan::Impl::getIndex()",
+            "Index %d is not valid.",
+            m_index);
+        return m_index;
+    }
+
+    // Mutable CableSpanData::Instance cache access.
+    CableSpanData::Instance& updDataInst(const State& state) const
+    {
+        return Value<CableSpanData::Instance>::updDowncast(
+            getSubsystem().updCacheEntry(state, m_indexDataInst));
+    }
+
+    // Mutable CableSpanData::Pos cache access.
+    CableSpanData::Pos& updDataPos(const State& state) const
+    {
+        return Value<CableSpanData::Pos>::updDowncast(
+            getSubsystem().updCacheEntry(state, m_indexDataPos));
+    }
+
+    // Mutable CableSpanData::Vel cache access.
+    CableSpanData::Vel& updDataVel(const State& state) const
+    {
+        return Value<CableSpanData::Vel>::updDowncast(
+            getSubsystem().updCacheEntry(state, m_indexDataVel));
+    }
+
+    //------------------------------------------------------------------------------
+    //  Computing cached data
+    //------------------------------------------------------------------------------
+
+    const MatrixWorkspace& calcDataInst(const State& state) const;
+
+    const CableSpanData::Pos& calcDataPos(const State& s) const
+    {
+        CableSpanData::Pos& dataPos = updDataPos(s);
+
+        dataPos.originPoint_G      = calcOriginPointInGround(s);
+        dataPos.terminationPoint_G = calcTerminationPointInGround(s);
+
+        // Helper for computing the total cable length.
+        auto calcTotalCableLength =
+            [&](const std::vector<LineSegment>& lines) -> Real
+        {
+            Real totalCableLength = 0.;
+            // Add length of each line segment.
+            for (const LineSegment& line : lines) {
+                totalCableLength += line.length;
+            }
+            // Add length of each curve segment.
+            forEachActiveCurveSegment(
+                s,
+                [&](const CurveSegment& curve)
+                {
+                    // Update the total cable length.
+                    totalCableLength += curve.getDataInst(s).arcLength;
+                });
+            return totalCableLength;
+        };
+
+        // Computing the CableSpan's path is done iteratively by computing
+        // corrections for the CurveSegments until the pathErrorVector is small
+        // enough.
+        dataPos.loopIter = 0;
+        while (true) {
```

Comment:
Please just have upper limits on these loops, or actually just put the condition in the loop (`while loopIter < getParameters().solverMaxIterations`) - seeing `while (true)` in source code is just a recipe for anxiety